cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java

```diff
@@ -162,7 +167,25 @@ public int hashCode() {
       }
 
       LOG.debug("New stream file created at {}", eventFile);
-      return new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      final StreamDataFileWriter streamDataFileWriter =
+        new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      timer.schedule(new TimerTask() {
+        @Override
+        public void run() {
+          try {
+            // NOTE: this doesn't remove it from PartitionedFileWriter map of writers
+            // it will get removed from there when the next event comes
+            streamDataFileWriter.close();
+          } catch (IOException e) {
+            Throwables.propagate(e);
+          }
+        }
+        // add a 5 minute buffer, to ensure that all events for this time partition are enqueued to the writer
+        // before closing it
+      }, partitionDuration + TimeUnit.MINUTES.toMillis(5));
```

Comment:
Enqueuing events to the writer (after computing a time for that event) is a very quick process.
Much of the delay in writing will be after its already enqueued to that writer.